automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala

```diff
@@ -259,222 +198,70 @@ class LeonardoSpec extends FreeSpec with Matchers with Eventually with ParallelT
-          }
-        }
-      }
+  def withFileUpload[T](cluster: Cluster, file: File)(testCode: NotebooksListPage => T)(implicit webDriver: WebDriver, token: AuthToken): T = {
+    withNotebooksListPage(cluster) { notebooksListPage =>
+      notebooksListPage.upload(file)
+      testCode(notebooksListPage)
     }
+  }
 
-    "should create a cluster in a different billing project and put the pet's credentials on the cluster" in withWebDriver { implicit driver =>
-      withNewBillingProject { newProject =>
-
-        implicit val token = ronAuthToken
-        // Pre-conditions: pet service account exists in this Google project and in Sam
-        val (petName, petEmail) = getAndVerifyPet(newProject)
-
-        // Create a cluster
-
-        withNewCluster(newProject) { cluster =>
-          // cluster should have been created with the pet service account
-          cluster.serviceAccountInfo.clusterServiceAccount shouldBe Some(petEmail)
-          cluster.serviceAccountInfo.notebookServiceAccount shouldBe None
-
-          withNewNotebook(cluster) { notebookPage =>
-            // should not have notebook credentials because Leo is not configured to use a notebook service account
-            verifyNoNotebookCredentials(notebookPage)
-          }
-        }
-
-        // Post-conditions: pet should still exist in this Google project
-
-        val googlePetEmail2 = googleIamDAO.findServiceAccount(newProject, petName).futureValue.map(_.email)
-        googlePetEmail2 shouldBe Some(petEmail)
-      }
+  def withNotebookUpload[T](cluster: Cluster, file: File)(testCode: NotebookPage => T)(implicit webDriver: WebDriver, token: AuthToken): T = {
+    withFileUpload(cluster, file) { notebooksListPage =>
+      val notebookPage = notebooksListPage.openNotebook(file)
+      val result = testCode(notebookPage)
+      notebookPage.terminateKernelAndExit()
+      result
     }
+  }
 
-    "should do cross domain cookie auth" in withWebDriver { implicit driver =>
-      withNewBillingProject { project =>
-        implicit val token = ronAuthToken
-        withNewCluster(project) { cluster =>
-          withDummyClientPage(cluster) { dummyClientPage =>
-            // opens the notebook list page without setting a cookie
-            val notebooksListPage = dummyClientPage.openNotebook
-            val notebookPage = notebooksListPage.newNotebook
-            // execute some cells to make sure it works
-            notebookPage.executeCell("1+1") shouldBe Some("2")
-            notebookPage.executeCell("2*3") shouldBe Some("6")
-            notebookPage.executeCell("""print 'Hello Notebook!'""") shouldBe Some("Hello Notebook!")
-          }
-        }
-      }
+  def withNewNotebook[T](cluster: Cluster)(testCode: NotebookPage => T)(implicit webDriver: WebDriver, token: AuthToken): T = {
+    withNotebooksListPage(cluster) { notebooksListPage =>
+      val notebookPage = notebooksListPage.newNotebook
+      val result = testCode(notebookPage)
```

Comment:
Could `testCode` throw an exception, causing you to not terminate the kernel? Wondering if you need something like this:
```
val testResult: Try[T] = Try {
  testCode(notebookPage)
}
notebookPage.terminateKernelAndExit()
result
```

Reply:
good catch.  I'll take this opportunity to `with`ify it too.